tensorstore/kvstore/s3/credentials/file_credential_provider.cc

```diff
@@ -0,0 +1,147 @@
+
+#include "tensorstore/kvstore/s3/credentials/file_credential_provider.h"
+
+#include <fstream>
+#include <optional>
+#include <string>
+#include <string_view>
+
+#include "absl/log/absl_log.h"
+#include "absl/status/status.h"
+#include "absl/strings/ascii.h"
+#include "absl/strings/str_cat.h"
+#include "absl/time/time.h"
+#include "tensorstore/internal/env.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/util/result.h"
+
+using ::tensorstore::internal::GetEnv;
+using ::tensorstore::internal::JoinPath;
+
+namespace tensorstore {
+namespace internal_kvstore_s3 {
+
+namespace {
+
+// Credentials file environment variable
+static constexpr char kEnvAwsCredentialsFile[] = "AWS_SHARED_CREDENTIALS_FILE";
+// Default path to the AWS credentials file, relative to the home folder
+static constexpr char kDefaultAwsCredentialsFilePath[] = ".aws/credentials";
+// AWS user identifier
+static constexpr char kCfgAwsAccessKeyId[] = "aws_access_key_id";
+// AWS user password
+static constexpr char kCfgAwsSecretAccessKeyId[] = "aws_secret_access_key";
+// AWS session token
+static constexpr char kCfgAwsSessionToken[] = "aws_session_token";
+// Discover AWS profile in environment variables
+static constexpr char kEnvAwsProfile[] = "AWS_PROFILE";
+// Default profile
+static constexpr char kDefaultProfile[] = "default";
+
+Result<std::string> GetAwsCredentialsFileName() {
+  std::string result;
+
+  if (auto credentials_file = GetEnv(kEnvAwsCredentialsFile);
+      credentials_file) {
+    result = *credentials_file;
+  } else {
+    if (auto home_dir = GetEnv("HOME"); home_dir) {
+      result = JoinPath(*home_dir, kDefaultAwsCredentialsFilePath);
+    } else {
+      return absl::NotFoundError("Could not read $HOME");
+    }
+  }
+
+  if (auto fstream = std::ifstream(result.c_str()); !fstream.good()) {
+    return absl::NotFoundError(
+        absl::StrCat("Could not find the credentials file at "
+                     "location [",
+                     result, "]"));
+  }
+
+  return result;
+}
+
+}  // namespace
+
+/// https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html#cli-configure-files-format
+Result<AwsCredentials> FileCredentialProvider::GetCredentials() {
+  TENSORSTORE_ASSIGN_OR_RETURN(auto filename, ([&]() -> Result<std::string> {
+                                 if (!filename_.empty()) return filename_;
+                                 return GetAwsCredentialsFileName();
+                               }()));
+
+  std::ifstream ifs(filename.c_str());
+  if (!ifs) {
+    return absl::NotFoundError(
+        absl::StrCat("Could not open credentials file [", filename, "]"));
+  }
+
+  auto profile = !profile_.empty()
+                     ? std::string(profile_)
+                     : GetEnv(kEnvAwsProfile).value_or(kDefaultProfile);
+
+  auto credentials = AwsCredentials{};
+  auto section_name = std::string{};
+  auto line = std::string{};
+  auto profile_found = false;
+
+  while (std::getline(ifs, line)) {
```

Comment:
We have used riegeli for a lot of our common file io. In this case it would look like this:

  riegeli::FileReader reader(filename);
  if (!reader.ok()) {
    return reader.status();
  }

  std::string line;
  while (riegeli::ReadLine(reader, line)) {
    // current loop
  }

... though consider this informational, not required at the moment.